### ISSUE TITLE
Bowling Game

### ISSUE THREAD

**Issue by pixelmike** [schteppe/cannon.js#172]
Hi schteppe!

Yeah I know, it's probably one of the most obvious things to try with a physics engine, but it made for a quick and fun mobile-friendly game. Thought you might get a kick out of it:

http://www.spacejack.ca/bowling/

(You can use mobile Chrome/Safari "Add to homescreen" feature to get an icon on the home screen that plays fullscreen like an app.)

It runs pretty well on my Nexus 5. Though I notice on the first throw, the game slows down quite a bit when the ball hits the pins. After that it's a lot smoother for subsequent throws. I guess the VM is working on optimizing those code paths the first time around. I suppose I could throw an invisible ball at some invisible pins in the background before they start or something. Any other ideas on making this smoother for the first collisions?

One minor thing about the docs I thought I'd point out: In the docs for the Body class constructor, the "material" option is not listed. Took me a bit of searching through the demos to find out how to apply materials to bodies.

One other thing that might be nice to have described is how to "stop" an object. To reset standing pins after a throw that may be slightly rocking, I used:

```
body.position.copy(pos);
body.quaternion.copy(quat);
body.velocity.set(0,0,0);
body.angularVelocity.set(0,0,0);
//body.intertia.set(0,0,0); // does anything?
body.vlambda.set(0,0,0);
body.wlambda.set(0,0,0);
```

Nice work on the trimesh recently, that looks very cool!

**Comment by schteppe** [comment]
Sorry for the late answer, should really try to fix issues regularly...

I think that the first-throw slowdown is because of JITing and allocation of new objects. Cannon uses object pools to make the garbage collector happy, but these pools have to be filled on the fly when the number of contacts and other things in the scene increase. The solution for this would be to run a few timesteps with many objects in contact. Or some sort of .prefillObjectPools() method could be implemented in cannon that does the job. I'll think about that.

Resetting the pins should just be to set position and velocities. .vlambda and .wlambda are set to zero by the solver each step, so setting them should not be necessary. You can try and set inertia, but to update it you need to do .updateMassProperties() too.
Reposition the pins so they are slightly above the ground. If you teleport them to a new position where they are in contact with something, you will get unexpected results.

I really like your game, it's got a snappy and polished feel :D

**Comment by pixelmike** [comment]
Yes, I tried optimizing a few things since I wrote that. I did end up running the sim invisibly for a few seconds (crash the ball into the pins) when the main menu appears. This helped the first play quite a bit on mobile.

A couple of other things: 1) When the ball isn't rolling, there is no animation/physics update. 2) When the ball is first thrown, the pins are all removed from the world and aren't added until the ball gets very close to the first pin. That all improved things on mobile quite a bit, except for the worst case where 10 pins are all colliding, but I guess that is simply a heavy bit of physics that can't be avoided.

Thanks for the feedback!

**Comment by schteppe** [comment]
I think you could get use of the body sleeping feature, then the pins would be deactivated when they stand still for a while and wake up again when they are hit.

What shapes are the pins? Convexes? If you use box, cylinder and/or convex you will probably see a nice performance improvement if you get the very latest build.

Thank YOU for the feedback :)

**Comment by schteppe** [comment]
By the way, I've thrown together a [Cannon.js debug renderer for three.js](https://github.com/schteppe/cannon.js/tree/master/tools/threejs). If you've got time over, would you like to test it? (you only have to add 2 lines of code)

**Comment by pixelmike** [comment]
The pins are compound objects - a tall, thin box with a small sphere near the top and a larger sphere near the bottom. Seems to simulate a pin shape pretty well. :)

I think I tried sleep, but one issue I had is that with a variable framerate (mobile), the pins would sometimes start rocking on their own (and I did try spawning the pins a bit higher than the ground.) I played around with slicing long frames into multiple steps, but if I made the step too small, there would be too many physics updates per frame on mobile. I eventually compromised with 20ms (50fps) maximum step size, however the pins still sometimes "rock" at that rate. The easy solution was to remove them from the world until the ball was about to collide.

I'll test the debug renderer when I get a chance. It looks very helpful. :)
